velox/exec/HashTable.cpp

```diff
@@ -614,24 +619,27 @@ void HashTable<ignoreNullKeys>::clear() {
 
 template <bool ignoreNullKeys>
 void HashTable<ignoreNullKeys>::checkSize(int32_t numNew) {
+  const int64_t numDistincts = numNew + numDistinct_;
   if (!table_ || !size_) {
     // Initial guess of cardinality is double the first input batch or at
     // least 2K entries.
     // numDistinct_ is non-0 when switching from HashMode::kArray to regular
     // hashing.
     auto newSize = std::max(
         (uint64_t)2048, bits::nextPowerOfTwo(numNew * 2 + numDistinct_));
-    if (numNew + numDistinct_ > rehashSize(newSize)) {
+    if (numDistincts > rehashSize(newSize)) {
       newSize *= 2;
     }
     allocateTables(newSize);
     if (numDistinct_) {
       rehash();
     }
-  } else if (numNew + numDistinct_ > rehashSize()) {
+  } else if (numDistincts > size_ || numDistincts > rehashSize()) {
     auto newSize = bits::nextPowerOfTwo(size_ + numNew);
     allocateTables(newSize);
     rehash();
+  } else if (numDistincts > rehashSize(effectiveSize())) {
```

Comment:
How about changing `rehashSize()` to return `rehashSize(size_ - numTombstoneRows_)`?  Then you don't need this extra branch.